packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -1031,6 +1044,21 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         options: CrawlerAddRequestsOptions = {},
     ): Promise<CrawlerAddRequestsResult> {
         const requestQueue = await this.getRequestQueue();
+
+        if (this.respectRobotsTxtFile) {
+            const allowedRequests: (string | Source)[] = [];
+
+            for (const request of requests) {
+                const url = typeof request === 'string' ? request : request.url!;
+
+                if (await this.isAllowedBasedOnRobotsTxtFile(url)) {
+                    allowedRequests.push(request);
+                }
+            }
```

Comment:
random note: `addRequests()` is called on the first argument of `crawler.run()`.

If we're planning to make this default in 4.0, it would mean that the crawler spends the first few seconds loading the `robots.txt` files and only then starts processing the actual URLs.

Probably no biggie for the majority of users (so IMO no need for action now), but for people crawling loads of URLs from different domains, this might cause a surprise (no logging, so the crawler might seem "stuck").

Reply:
I hope it won't be seconds, since the request is fired without a proxy and the file will/should be really small. But we can always adjust, I am sure we'll learn about all those edge cases once we start adopting this in WCC.

Reply:
We could group these by domain (so we don't hit the same file multiple times) and then do this in concurrent batches. Maybe overkill for now but I would add at least comment with this idea because in some cases like 5000 random domains on startUrls, this might stall for few minutes

Reply:
> group these by domain (so we don't hit the same file multiple times) and then do this in concurrent batches

Alternatively, we can just put them all in the RQ and only skip them once we actually get to them. It might put a bigger load on the RQ, but it will allow us to interleave (possibly) good requests with blocked ones, minimizing the "time to first scraped result" (not having to wait for possibly unrelated `robots.txt` files to load).

Agreed with overkill for now, agreed with adding a comment with some of those ideas :)